io/vdma.cpp

```diff
@@ -0,0 +1,246 @@
+            close(shm_fd_);
+            return;
+        }
+
+        nbuffer_ = shm_size_ / unit_;
+        used_mark_.resize(nbuffer_);
+        buffers_.resize(nbuffer_);
+        LOG_INFO("vDMABufferAllocator: shm_begin_ptr=", shm_begin_ptr_, ", nbuffer=", nbuffer_);
+        LOG_INFO("vDMABufferAllocator: used_mark.size=", used_mark_.size(), ", buffers.size=", buffers_.size());
+        
+        for (int i=0; i<nbuffer_; i++) {
+            buffers_[i] = new vDMABufferImpl(i, shm_begin_ptr_ + i * unit_, unit_, vDMABufferType::kSharedMem);
+        }
+        LOG_INFO("vDMABufferAllocator: complete");
+    }
+
+    ~vDMABufferAllocator() {
+        SCOPED_LOCK(mutex_);
+        for (int i=0; i<nbuffer_; i++) {
+            vDMABufferImpl* buf = buffers_[i];
+            if (buf) delete buf;
+        }
+    }
+
+    vDMABuffer* alloc_one() {
+        SCOPED_LOCK(mutex_);
+        for (int i=0; i<nbuffer_; i++) {
+            if (!used_mark_[i]) {
+                used_mark_[i] = true;
+                return buffers_[i];
+            }
+        }
+        return nullptr;
+    }
+
+    int free_one(vDMABuffer* buf) {
+        SCOPED_LOCK(mutex_);
+        used_mark_[reinterpret_cast<vDMABufferImpl*>(buf)->idx()] = false;
+        return 0;
+    }
+
+    size_t unit() const { return unit_; }
+    
+private:
+    std::string shm_name_;
+    int shm_fd_;
+    size_t shm_size_;
+    char* shm_begin_ptr_;
+    
+    size_t unit_;
+    int nbuffer_;
+    const int max_retry_ = 10;
+
+    photon::mutex mutex_;
+    std::vector<vDMABufferImpl*> buffers_;
+    std::vector<bool> used_mark_;
+};
+
+class vDMATargetImpl : public vDMATarget {
+public:
+    vDMATargetImpl(std::string shm_name, size_t shm_size, size_t unit) {
+        allocator_ = new vDMABufferAllocator(shm_name, shm_size, unit);
+    }
+
+    ~vDMATargetImpl() {
+        if (allocator_) {
+            delete allocator_;
+        }
+    }
+
+    vDMABuffer* alloc(size_t size) override {
+        if (size != allocator_->unit()) {
+            LOG_ERROR("current allocator only support ", allocator_->unit(), ", you ", size);
+            return nullptr;
+        }
+
+        vDMABuffer* buf = nullptr;
+        buf = allocator_->alloc_one();
+
+        int retry_count = 0;
+        while (!buf && retry_count < max_retry_) {
+            thread_yield();
+            buf = allocator_->alloc_one();
+            retry_count++;
+        }
+
+        return buf;
+    }
+
+    int dealloc(vDMABuffer* buf) override {
+        return allocator_->free_one(buf);
+    }
+
+    vDMABuffer* register_memory(char* buf, size_t size) override {
+        LOG_WARN("register_memory is empty function");
+        return nullptr;
+    }
+
+    int unregister_memory(vDMABuffer* vbuf) override {
+        LOG_WARN("unregister_memory is empty function");
+        return -1;
+    }
+
+private:
+    vDMABufferAllocator* allocator_;
+    const int max_retry_ = 10000;
```

Comment:
use constexpr or const static instead.

now as a const member, `max_retry_` will actually consume 4bytes in every vDMABuffer object